### PR TITLE
Use unique namespace names for central deployments

### PR DESCRIPTION
**Pull request by kovayur**
## Description
Use unique namespace names for central deployments.

## Checklist (Definition of Done)

- [x] Unit and integration tests added
- [ ] Documentation added if necessary
- [ ] CI and all relevant tests are passing

## Test manual
1. Add yourself in the `quota-management-list-configuration.yaml`
2. Create a few instances using public API
3. Make sure that unique namespaces were created by the fleetshard

```
# To run tests locally run:
make db/teardown db/setup db/migrate
make ocm/setup OCM_OFFLINE_TOKEN=<ocm-offline-token> OCM_ENV=development
make verify lint binary test test/integration
```

### CODE REVIEW

File: internal/dinosaur/pkg/services/util.go [hunk 1]
```diff
@@ -19,6 +19,11 @@ const (
 	appendChar                = "a"
 )
 
+const dns1123LabelFmt string = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+const dns1123LabelErrMsg string = "a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character"
```
Comment (SimonBaeumer) [inline internal/dinosaur/pkg/services/util.go:23]: nit: Use instead
```
const (
   ...
)
```

File: internal/dinosaur/pkg/services/util.go [hunk 2]
```diff
@@ -63,3 +68,12 @@ func replaceHostSpecialChar(name string) (string, error) {
 
 	return replacedName, nil
 }
+
+// formatNamespace adds the rhacs prefix to the namespace name and performs the necessary validation and formatting to comply with RFC1123 that namespace names must follow.
+func formatNamespace(text string) (string, error) {
+	if !dns1123LabelRegexp.MatchString(text) {
+		return "", fmt.Errorf("invalid namespace %s: %s", text, validation.RegexError(dns1123LabelErrMsg, dns1123LabelFmt, "my-name", "123-abc"))
```
Comment (SimonBaeumer) [inline internal/dinosaur/pkg/services/util.go:75]: Are `my-name` and `123-abc` placeholders?

File: internal/dinosaur/pkg/services/util_test.go [hunk 3]
```diff
@@ -352,3 +353,40 @@ func Test_contains(t *testing.T) {
 		})
 	}
 }
+
+func Test_formatNamespace(t *testing.T) {
+	type testCase struct {
+		id        string
+		namespace string
+		error     bool
```
Comment (SimonBaeumer) [inline internal/dinosaur/pkg/services/util_test.go:361]: I prefer being explicit with properties in test. In this case for example not obvious which of these properties is the expected result.

```suggestion
		expectError     bool
```

File: internal/dinosaur/pkg/services/util.go [hunk 4]
```diff
@@ -19,6 +19,11 @@ const (
 	appendChar                = "a"
 )
 
+const dns1123LabelFmt string = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+const dns1123LabelErrMsg string = "a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character"
+
+var dns1123LabelRegexp = regexp.MustCompile("^" + dns1123LabelFmt + "$")
```
Comment (SimonBaeumer) [inline internal/dinosaur/pkg/services/util.go:25]: nit:
```
var (
  ...
)
```

File: internal/dinosaur/pkg/services/util.go [hunk 5]
```diff
@@ -63,3 +68,12 @@ func replaceHostSpecialChar(name string) (string, error) {
 
 	return replacedName, nil
 }
+
+// formatNamespace adds the rhacs prefix to the namespace name and performs the necessary validation and formatting to comply with RFC1123 that namespace names must follow.
+func formatNamespace(text string) (string, error) {
+	if !dns1123LabelRegexp.MatchString(text) {
+		return "", fmt.Errorf("invalid namespace %s: %s", text, validation.RegexError(dns1123LabelErrMsg, dns1123LabelFmt, "my-name", "123-abc"))
+	}
+	truncatedNamespace := truncateString("rhacs-"+text, validation.LabelValueMaxLength)
```
Comment (SimonBaeumer) [inline internal/dinosaur/pkg/services/util.go:77]: WDYTH about adding actual instance or customer names to namespaces for easier navigation?
(More detailed data of could be added to the namespace labels anyway)

**Review by SimonBaeumer — CHANGES_REQUESTED** [review]

File: internal/dinosaur/pkg/services/util.go (hunk 1 above)
Comment (kovayur) [reply internal/dinosaur/pkg/services/util.go]: Fixed.

File: internal/dinosaur/pkg/services/util.go (hunk 2 above)
Comment (kovayur) [reply internal/dinosaur/pkg/services/util.go:75]: These are examples of the correct format

File: internal/dinosaur/pkg/services/util_test.go (hunk 3 above)
Comment (kovayur) [reply internal/dinosaur/pkg/services/util_test.go]: Renamed as suggested.

**Pull request merged**